ui/dashboardApp/index.ts

```diff
@@ -75,12 +84,60 @@ async function main() {
     if (!routing.isLocationMatchPrefix(auth.signInRoute)) {
       if (!auth.getAuthTokenAsBearer()) {
         singleSpa.navigateToUrl('#' + auth.signInRoute)
-        return
       }
     }
   })
 
   singleSpa.start()
 }
 
-main()
+/////////////////////////////////////
+
+// check whether this is a portal page
+function checkPortal(): boolean {
+  return window.location.pathname.endsWith('/portal')
+}
+
+// check whether the url has 'iframe' query parameter
+function checkIframe(): boolean {
+  const hash = window.location.hash
+  const pos = hash.indexOf('?')
+  if (pos === -1) {
+    return false
+  }
+  let q = hash.substring(pos + 1)
+  const p = new URLSearchParams(q)
+  return p.get('iframe') !== null
+}
+
+function start() {
+  // non portal page
+  if (!checkPortal()) {
+    main({ hideNav: false, lang: '' })
+    return
+  }
+
+  // use another auth token key for portal page to avoid affect the non portal page
+  auth.setTokenKey(auth.portalTokenKey)
+
+  // portal page runs in iframe
+  if (checkIframe()) {
```

Comment:
We can make life simple by stipulating that the portal page always accepts config (or reset config) from `message` event and the other pages always loads config from localStorage (or if non-exist, using default config). In this way, we don't need to check the existence of `iframe` as well as supporting multiple token keys.

Reply:
Understand the whole logic finally and update it!